platform/darwin/src/MGLFeature.mm

```diff
@@ -5,14 +5,105 @@
+            });
+            return { (double)number.doubleValue };
+        } else if ([number compare:@(0)] == NSOrderedDescending ||
+                   [number compare:@(0)] == NSOrderedSame) {
+            // Positive integer or zero; use uint64_t per mbgl::Value definition.
+            // We use unsigned long long here to avoid any truncation.
+            return { (uint64_t)number.unsignedLongLongValue };
+        } else if ([number compare:@(0)] == NSOrderedAscending) {
+            // Negative integer; use int64_t per mbgl::Value definition.
+            // We use long long here to avoid any truncation.
+            return { (int64_t)number.longLongValue };
+        }
+    }
+    [NSException raise:@"Value not handled"
+                format:@"Can’t convert %s:%@ to mbgl::Value", [value objCType], value];
+    return { };
+}
+
++ (mbgl::PropertyMap)propertyMapForDictionary:(NSDictionary *)dictionary propertyMap:(mbgl::PropertyMap *)propertyMap {
+    for (NSString *key in dictionary.allKeys) {
+        if ([dictionary[key] isKindOfClass:[NSDictionary class]]) {
+            std::unordered_map<std::string, mbgl::Value> unorderedMap;
+            [self propertyMapForDictionary:dictionary[key] propertyMap:&unorderedMap];
+            (*propertyMap)[[key cStringUsingEncoding:NSUTF8StringEncoding]] = unorderedMap;
+        } else if ([dictionary[key] isKindOfClass:[NSArray class]]) {
+            NSArray *array = dictionary[key];
+            std::vector<mbgl::Value> vector;
+            (*propertyMap)[[key cStringUsingEncoding:NSUTF8StringEncoding]] = [self vectorForNSArray:array vector:&vector];
+        } else {
+            (*propertyMap)[[key cStringUsingEncoding:NSUTF8StringEncoding]] = [MGLFeatureUtility convertedValueWithValue:dictionary[key]];
+        }
+    }
+    
+    return *propertyMap;
+}
+
++ (std::vector<mbgl::Value>)vectorForNSArray:(NSArray *)array vector:(std::vector<mbgl::Value> *)vector {
+    (*vector).reserve(array.count);
+    for (id value in array) {
+        if ([value isKindOfClass:[NSArray class]]) {
+            std::vector<mbgl::Value> innerVector;
+            [self vectorForNSArray:value vector:&innerVector];
+            (*vector).push_back(innerVector);
+        } else if ([value isKindOfClass:[NSDictionary class]]) {
+            mbgl::PropertyMap propertyMap;
+            [MGLFeatureUtility propertyMapForDictionary:value propertyMap:&propertyMap];
+            (*vector).push_back(propertyMap);
+        } else {
+            (*vector).push_back([MGLFeatureUtility convertedValueWithValue:value]);
+        }
+    }
+    
+    return *vector;
+}
+
++ (mbgl::Feature)mbglFeatureWithPropertyMapForFeature:(mbgl::Feature)feature mglAttributes:(NSDictionary *)attributes {
+    mbgl::PropertyMap propertyMap;
+    [MGLFeatureUtility propertyMapForDictionary:attributes propertyMap:&propertyMap];
```

Comment:
Would it be more convenient for NSDictionary to know how to turn itself into a PropertyMap?